### PR TITLE
Add annotations to System.Private.DataContractSerialization

### PR DESCRIPTION
**Pull request by tlakollo**
Enables AOT analyzer and adds RDC annotations

### CODE REVIEW

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs [hunk 1]
```diff
@@ -65,12 +68,14 @@ private sealed class CriticalHelper
...
 
             [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+            [RequiresDynamicCode(DataContract.SerializerAOTWarning)]
             private static XmlFormatClassReaderDelegate CreateReflectionXmlClassReader(ClassDataContract classContract)
             {
                 return new ReflectionXmlClassReader(classContract).ReflectionReadClass;
             }
 
             [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+            [RequiresDynamicCode(DataContract.SerializerAOTWarning)]
```
Comment (MichalStrehovsky) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs:78]: Is this one RDC even with the knowledge that the `DataContractSerializer.Option == SerializationOption.ReflectionOnly` codepath will always be taken when AOT compiling?

`DataContractSerializer.Option` basically controls whether we're going to take the ref.emit code paths.

Comment (tlakollo) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs:78]: Both implementations are flagged by the analyzer as annotated with RDC, so the option seems to not affect the end result. I tried to go depper on the ReflectionOnly call and at several points it calls multiple methods annotated with RDC

**Review by vitek-karas — APPROVED** [review]
I think this is a good start.
It might be worth filing an issue that even asking for reflection-only (no ref.emit) doesn't make it AOT friendly. Maybe it's not possible, but currently we don't know why.

**Comment by StephenMolloy** [comment]
Thanks for the heads up on this guys. We've been working on a huge PR for DCR for months and this has just caused a slew of conflict resolutions for that PR in our area at the last minute before we were set to checkin.

**Comment by StephenMolloy** [comment]
Actually, after looking at all the conflicts that need resolving, there is so much that has changed here, I don't want to try and figure out how to keep your annotations correct in the new code. I'm reverting this PR.

**Comment by tlakollo** [comment]
@StephenMolloy can you add a reference to the big PR you mentioned to know when is safe to reenable the analyzer?

**Comment by StephenMolloy** [comment]
Sure. [#71752](https://github.com/dotnet/runtime/pull/71752) Once I can merge the revert, I'll merge that PR and ping back here.

**Comment by StephenMolloy** [comment]
#71752 has been merged. Thanks for understanding.

**Pull request merged**